pkg/client/client.go

```diff
@@ -0,0 +1,14 @@
+package client
+
+// Facotry creates new clients
+type Facotry interface {
+	NewOvirtClient(dataMap map[string]string) (VMClient, error)
+}
+
+// VMClient provides interface how source virtual machines should be fetched
+type VMClient interface {
```

Comment:
I am not sure that it needs to be so generic - other providers might need other operations and the operations are executed only by the concrete providers.

What do you think about creating provider factory, that would assemble providers with what they need instead? Is control over vm client needed in "general" packages? All that can be mocked on the client level can be mocked on provider (from controller perspective). I think vm client should be an implementation detail of each provider.

Generally I think that layer of code should only have access to abstration over its lower level if it needs to be both extensible and loosely-coupled. In this case we give controller access to abstraction over code two levels below.

Reply:
Interesting idea, will explore.

Reply:
I thought about it how to do it well and it seems that provider factory won't bring a lot of value. We still would need a way to detach client and provider life cycles. We want to use the client when it is needed as specified in issue https://github.com/kubevirt/vm-import-operator/issues/64. I will move it to `OvirtProvider`.

Reply:
Do you plan to use it directly from the controller? Won't provider lifecycle methods (maybe split/refactored) be sufficient?

Reply:
The idea is to have a factory which we can use to create client. In a unit test we need to have a way to mock the client. Provider methods are not sufficient for it.

Reply:
:+1: